Decoupling the transport

jneem opened the pull request:
This isn't done yet, but if you have any feedback on the general direction they I'd be happy to hear it.

Right now, this removes all of the network-related code and adds three main entry points (one each for client, server, and client+server).

It also does various other random things, like changing the way it gets back responses from the Service. I'm not really convinced by the ReturnChannel stuff; so far, I'm just playing around with it.

jneem committed:
Rip out TLS support.

jneem committed:
Remove all net code, and make messages public.

jneem committed:
Here's an attempt to update the ping_pong example but it doesn't work yet -- it seems like the Client is getting dropped, and that causes everything to shut down prematurely.

jneem committed:
Get all the basic examples working.

jneem committed:
Run clippy, and remove non-compiling examples.

jneem committed:
Remove some type parameters and start documenting.

little-dude commented:
Hi @jneem,

Thanks a lot for the PR. I'm currently traveling, but I'll try to look at it tomorrow.

little-dude commented on src/endpoint.rs:
```diff
@@ -13,90 +13,174 @@ use message::{Message, Notification, Request};
+#[derive(Debug)]
+pub struct ReturnChannel {
+    id: u32,
+    sender: mpsc::UnboundedSender<(u32, Result<Value, Value>)>,
```
Since `send()` consumes the `ReturnChannel`, shouldn't this be a `oneshot` channel?

little-dude commented on src/endpoint.rs:
```diff
@@ -13,90 +13,174 @@ use message::{Message, Notification, Request};
+    /// because it complicated the easy case where the response is immediately available. This
+    /// "return channel" method also has the (potential) advantage of allowing the service to be in
+    /// control of spawning the new task.
+    fn handle_request(&mut self, method: &str, params: &[Value], return_channel: ReturnChannel);
```
I'm not sure how I feel about this. On one hand, not returning a future has all the advantages you list, and also has the benefit of simplifying error handling. But on the other hand it's exposing an implementation detail to the users (they have to understand what this `return_channel` does, and use it correctly). But to be honest I lack experience designing rust APIs (I made `rmp-rpc` to learn about Tokio and async programming in the first place), I can't decide which way I prefer. @tylerwhall do you have an opinion about this?

jneem replied:
So I looked at [tarpc](https://github.com/google/tarpc) yesterday, and they have a nicer approach: the `Service` trait would come with a `type Fut: IntoFuture<Item=Value, Error=Value>` associated type, which is then the return value of all the methods. That way, if you just want to be synchronous then you can set `Fut` to `Result`.

little-dude replied:
That sounds like the best of both world, indeed :)

Now that I take a closer look to this, I realize your current changes remove the possiblity to express a failure to handle a request or a notification, but that's fine, because anyway I [currently panic](https://github.com/little-dude/rmp-rpc/blob/0dd5e8e44b2f56ba58725b61f58ecd96cf64fa32/src/endpoint.rs#L70) if `handle_request` or `handle_notification` returns an error, which is way worse.

jneem replied:
Well, it seemed to me that the only reasonable way to deal with an error from within `handle_request` is either to panic (if the error represents a real fatal bug) or send an error message back to the client (if the error is recoverable). In the panic case, there's no need to bubble the error up before panicking.

little-dude reviewed (commented):
Thanks a lot for all this work! I like the idea of abstracting over the transport layer instead of hardcoding TCP. Also, thanks for adding all this documentation and comments.

These are breaking changes obviously, so we'll have to publish a 0.2.

@tylerwhall you're the only person I know of that might be using this crate, so if you have an opinion about these changes please let us know :) That removes the TLS support you added, but since you can know use the transport layer you want, it should be easy enough for users to use a TLS stream.

jneem committed:
Remove ReturnChannel.

jneem committed:
Avoid spawning a worker future, if possible.

little-dude commented:
@jneem are you done with this PR? I like the changes a lot!

I have a question: why do you write

```rust
type Future = <F as IntoFuture>::Future;
type Item = <F as IntoFuture>::Item;
type Error = <F as IntoFuture>::Error;
```

instead of

```rust
type Future = F::Future;
type Item = F::Item;
type Error = F::Error;
```

I've come accross this syntax a few times in other projects but I don't really understand what it does. Do you mind explaining what the difference is?

little-dude reviewed (approved):
This is a big improvement, thanks a lot :+1:

Commit:
fix clippy warning in server example

jneem commented:
Glad you like it :) I'd just like to finish a few more TODOs (mainly documentation) and then I think I'm done.

The `<F as IntoFuture>::Future` thing is because `F::Future` would be ambiguous: `F` implements multiple traits that have a `Future` associated type, so which one does it mean? Usually, I only ever introduce this as a response to a compiler error...

Commit:
tweak Service::new() signature

Do not pass the tokio::reactor::Handle() by reference.

This does not change much because anyway the service needs to own the
handle and ends up cloning it. It also makes it to write small helper
function that create endpoint like:

> type SessionFuture<H> =
>     Box<Future<Item = Endpoint<H, TcpStream>, Error = ()>>;
>
> pub fn new_tcp<H>(addr: &SocketAddr, handler: H, handle: Handle)
>     -> SessionFuture<H>
> where
>     H: ServiceWithClient + 'static,
> {
>     let endpoint = TcpStream::connect(&addr, &handle)
>         .map(|stream| Endpoint::new(stream, handler, handle))
>         .map_err(|_| ());
>     Box::new(endpoint)
> }

If we just pass a reference do Endpoint::new, rustc complains that it
does not live for the 'static lifetime.

jneem committed:
Add docs, and remove ClientEndpoint.

jneem commented:
This latest commit also introduces an API change, because as I was documenting ClientEndpoint it occurred to me that the API should be simpler. The point is that in `Endpoint`, there are some choices to be made about how long the `Endpoint` future should run, and so it makes sense to return that future to the caller. But for `ClientEndpoint`, it seems to me that there's really only one sensible choice, so we should just spawn it ourselves.

tylerwhall commented:
I'm much in favor of decoupling the transport. I actually made the TLS changes so I could build without TLS support. I haven't looked over this in detail, but I think it's good at a high level.

One comment on the Service trait: I would get rid of the NotificationFuture. I think if the service needs to perform async work in response to a notify, it could spawn its work itself however it chooses. The Service can get a tokio handle when it's created if that's what the app needs.

Longer-term, I'd like a way for the service to serialize directly to the output stream, avoiding the dynamic Value type. That might have to be another API breaking change down the road (or a new type of Service trait). My use is a bare-metal microcontroller environment, so I'm going to want to avoid the unbounded dynamic allocations of a Value tree. Possibly the Value-based Service could be implemented in terms of a lower-level trait that deals with raw msgpack streams or serde Serialize objects. Just something to think about.

Thanks for the work!

jneem commented:
Hm, that's a good point about NotificationFuture. I'm not completely convinced, though, because I'm a little bit inclined to have it be symmetric with RequestFuture.

I also have a slightly better, but still half-baked, motivation for keeping the NotificationFuture: the current design is vulnerable to DoS if a client makes a lot of quick requests, each of which spawn a future that takes some resources. I think the solution is for the server to pause reading from the stream until some of the futures finish. I'm not sure what the best API for that is (I have an experiment [here](https://github.com/jneem/rmp-rpc/tree/job_limit)), but I expect that having similar paths for spawning jobs in handle_request and handle_notification will make it easier.

I agree that it would be nice to avoid the dynamic Value type, and I'll open another issue for that.

little-dude commented:
I don't have a strong opinion about `NotificationFuture` myself.

> I agree that it would be nice to avoid the dynamic Value type, and I'll open another issue for that.

:+1: 

--------------

FYI I'll be offline until Monday, so I won't be able to comment anymore or merge until then.

Commit:
fix links in documentation

little-dude commented:
PS: not sure if you saw it, I opened a PR on your branch @jneem https://github.com/jneem/rmp-rpc/pull/1. Only minor clippy and documentation things, and one small change in the signature of `Service::new()`.

Commit:
update README

Commit:
simplify CI now that TLS feature is gone

jneem committed:
Merge pull request #1 from little-dude/clippy-fix

minor tweaks

tylerwhall commented:
My counterpoint about `NotificationFuture` would be that it makes sense for methods and notifications to not be symmetric because methods, by nature, have a return value while notifications do not. But it's not a big deal to return an empty result future either, since you poll it once before spawning. It just adds an associated type that people that don't need it will have to understand and specify.

jneem committed:
Remove NotificationFuture.

jneem commented:
Ok, I think I buy that. Also, I've been playing around with job limiting
and I don't think it should be "our" responsibility anyway.

On Thu, Mar 15, 2018 at 2:40 PM, Tyler Hall <notifications@github.com>
wrote:

> My counterpoint about NotificationFuture would be that it makes sense for
> methods and notifications to not be symmetric because methods, by nature,
> have a return value while notifications do not. But it's not a big deal to
> return an empty result future either, since you poll it once before
> spawning. It just adds an associated type that people that don't need it
> will have to understand and specify.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/little-dude/rmp-rpc/pull/12#issuecomment-373498771>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABYYsXp46nuXh3nSpQ_hNLhPN3BCk8ZRks5tesOggaJpZM4Sk1sp>
> .
>

Commit:
document and re-export IntoStaticFuture

Commit:
fix ServiceWithClient example

little-dude commented:
I'm back earlier than expected, so let try to get this merged this weekend.
@jneem I opened another PR to make `IntoStaticFuture` public and fix the failing test: https://github.com/jneem/rmp-rpc/pull/2.

Are you still working on further changes?

jneem committed:
Merge pull request #2 from little-dude/into-static-future

document and re-export IntoStaticFuture

jneem committed:
Run clippy.

jneem commented:
Welcome back! I'm happy with this PR if you are.

little-dude commented:
I'm very happy with this yes, thank you again!
